resources/external_control.urscript

```diff
@@ -180,18 +180,66 @@ def quinticSplineRun(end_q, end_qd, end_qdd, time):
     local coefficients3 = (-20.0 * start_q + 20.0 * end_q - 3.0 * start_qdd * TIME2 + end_qdd * TIME2 - 12.0 * start_qd * time - 8.0 * end_qd * time) / (2.0 * pow(time, 3))
     local coefficients4 = (30.0 * start_q - 30.0 * end_q + 3.0 * start_qdd * TIME2 - 2.0 * end_qdd * TIME2 + 16.0 * start_qd * time + 14.0 * end_qd * time) / (2.0 * pow(time, 4))
     local coefficients5 = (-12.0 * start_q + 12.0 * end_q - start_qdd * TIME2 + end_qdd * TIME2 - 6.0 * start_qd * time - 6.0 * end_qd * time) / (2.0 * pow(time, 5))
-    jointSplineRun(coefficients1, coefficients2, coefficients3, coefficients4, coefficients5, time)
+    jointSplineRun(coefficients1, coefficients2, coefficients3, coefficients4, coefficients5, time, last_point)
   end
 end
 
-def jointSplineRun(coefficients1, coefficients2, coefficients3, coefficients4, coefficients5, splineTotalTravelTime):
-  # Initialize incremental time variable
+def jointSplineRun(coefficients1, coefficients2, coefficients3, coefficients4, coefficients5, splineTotalTravelTime, last_point):
+  # Initialize variables
   local splineTimerTraveled = 0.0
+  local step_time = get_steptime()
+  local scaling_factor = 1.0
+  local slowing_down = False
+  local slowing_down_time = 0.0
 
   # Interpolate the spline in hole time steps
   while (splineTotalTravelTime - splineTimerTraveled) > get_steptime():
-    splineTimerTraveled = splineTimerTraveled + get_steptime()
-    jointSplineStep(coefficients1, coefficients2, coefficients3, coefficients4, coefficients5, splineTimerTraveled, get_steptime())
+    local time_left = splineTotalTravelTime - splineTimerTraveled
+    local deceleration_time = 0.4189
+
+    # If the velocity is too large close to the end of the trajectory we scale the
+    # trajectory, such that we follow the positional part of the trajectory, but end with zero velocity.
+    if time_left <= deceleration_time and last_point == True:
+      if slowing_down == False:
+        local deceleration = 15
+        local qd = jointSplinePeek(coefficients1, coefficients2, coefficients3, coefficients4, coefficients5, splineTimerTraveled + get_steptime())
+        local x = deceleration_time - (time_left - get_steptime()) # we look one time step ahead
```

Comment:
is `time_left` guaranteed to be greater than step_time?

Reply:
Yes because this is inside the while loop which only runs for as long as the time_left is greater than the step time.